File: console/src-tauri/src/adapters/workspace_profile.rs
use lunar::adapters::meta::RequestMeta;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateWorkspaceProfile {
    pub first_name: String,
    pub last_name: String,
    pub profile_picture: Option<String>,
    pub meta: Option<RequestMeta>,
}

impl From<CreateWorkspaceProfile> for lunar::adapters::workspace_profiles::CreateWorkspaceProfile {
    fn from(p: CreateWorkspaceProfile) -> Self {
        Self {
            first_name: p.first_name,
            last_name: p.last_name,
            profile_picture: p.profile_picture,
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateWorkspaceProfile {
    pub first_name: Option<String>,
    pub last_name: Option<String>,
    pub profile_picture: Option<String>,
    pub meta: Option<RequestMeta>,
}

impl From<UpdateWorkspaceProfile> for lunar::adapters::workspace_profiles::UpdateWorkspaceProfile {
    fn from(p: UpdateWorkspaceProfile) -> Self {
        Self {
            first_name: p.first_name,
            last_name: p.last_name,
            profile_picture: p.profile_picture,
        }
    }
}

File: console/app/layers/shared/stores/workspace-profile.ts
import type {
  CreateWorkspaceProfile,
  UpdateWorkspaceProfile,
  WorkspaceProfiles
} from "lunar";
import { defineStore } from "pinia";
import { invoke } from "../utils/invoke";
import { getWorkspaceMeta } from "#imports";

export const useUserPreferenceStore = defineStore("user_preference_store", {
  state: () => ({
    preference: null as WorkspaceProfiles | null,
    loading: false,
  }),

  getters: {
    fullName: (state) =>
      state.preference
        ? `${state.preference.firstName} ${state.preference.lastName}`.trim()
        : "",
  },

  actions: {
    async fetchPreference() {
      this.loading = true;
      try {
        this.preference = await invoke<WorkspaceProfiles | null>(
          "get_workspace_profile",
          {
            meta: await getWorkspaceMeta(),
          },
        );
      } catch (error) {
        console.error("[preferences] failed to fetch", error);
      } finally {
        this.loading = false;
      }
    },

    async createPreference(
      payload: CreateWorkspaceProfile,
    ): Promise<WorkspaceProfiles> {
      const created = await invoke<WorkspaceProfiles>(
        "create_workspace_profile",
        {
          profile: payload,
          meta: await getWorkspaceMeta(),
        },
      );
      this.preference = created;
      return created;
    },

    async updatePreference(
      payload: Partial<UpdateWorkspaceProfile>,
    ): Promise<WorkspaceProfiles> {
      if (!this.preference) {
        return this.createPreference(payload as CreateWorkspaceProfile);
      }
      const updated = await invoke<WorkspaceProfiles>(
        "update_workspace_profile",
        {
          identifier: this.preference.identifier,
          profile: payload,
          meta: await getWorkspaceMeta(),
        },
      );
      this.preference = updated;
      return updated;
    },

    async fetchUnsynced() {
      try {
        const userPreferences = await invoke<WorkspaceProfiles[]>(
          "get_unsynced_workspace_profiles",
        );
        return userPreferences;
      } catch (error) {
        console.error("Error fetching unsynced user preferences:", error);
        return [];
      }
    },

    async syncUpstream() {},

    async clearQueue(identifiers: string[]) {
      await invoke("clear_synced_workspace_profiles", { identifiers });
    },
  },
  persist: true,
});
